[RFC] Documentation and comment fixes in kernel crate

vobst opened the pull request:
Hi all!

To get familiar with the Rust kernel infrastructure I read through the existing in-tree code. Along the way I fixed some trivial things in comments and documentation, as well as making some small improvements by adding doclinks. This PR only looks at the kernel crate.

- Nothing here touches actual code.
- I tested the changes by recompiling the documentation (adding `--document-private-items`), running `make rusttest` to (hopefully) run the doctests, and running `make rustfmtcheck` to check the formatting. (Also recompiled and ran the the kernel and smaple modules, but this actually should not matter.)

Since this would be my first upstream contribution I'd appreciate some upfront review before I submit to the mailing list. In particular:
- are the proposed changes OK?
- is the splitting into individual commits appropriate?
- are the commit messages OK?
- should I make one or two patch sets?

Thanks!

charmitro commented:
>     * are the commit messages OK?

Two notes in the case you send this upstream via mailing list.

1. The commits in github are from an unknown profile. You can fix it by [modifying your git config](https://git-scm.com/docs/git-config#Documentation/git-config.txt-username). 
2. You should add `Signed-off-by: ...` in the commits.

In general, if you fix your git config, signing the commits can be done automatically via `git commit -s`.

ojeda commented on rust/kernel/workqueue.rs:
```diff
@@ -414,7 +414,7 @@ impl<T: ?Sized, const ID: u64> Work<T, ID> {
 /// # Safety
 ///
-/// The [`OFFSET`] constant must be the offset of a field in Self of type [`Work<T, ID>`]. The methods on
+/// The [`OFFSET`] constant must be the offset of a field in `Self` of type [`Work<T, ID>`]. The methods on
```
Can this be an intra-doc link?

vobst replied:
I think it can't as it refers to the type implementing the trait.

ojeda commented on rust/kernel/workqueue.rs:
```diff
@@ -428,7 +428,7 @@ pub unsafe trait HasWork<T, const ID: u64 = 0> {
     /// Returns the offset of the [`Work<T, ID>`] field.
     ///
-    /// This method exists because the [`OFFSET`] constant cannot be accessed if the type is not Sized.
+    /// This method exists because the [`OFFSET`] constant cannot be accessed if the type is not `Sized`.
```
Can this be an intra-doc link?

ojeda commented on rust/kernel/ioctl.rs:
```diff
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: GPL-2.0
 
-//! ioctl() number definitions
+//! ioctl() number definitions.
```
This could use Markdown, i.e. backquotes for `ioctl()` (in the commit that does that kind of change, assuming you didn't do it already).

ojeda commented on rust/kernel/print.rs:
```diff
@@ -189,7 +189,7 @@ macro_rules! print_macro (
 // imported in the prelude due to [1]. So, for the moment, we just write them
 // manually, like in the C side; while keeping most of the logic in another
-// macro, i.e. [`print_macro`].
+// macro, i.e. `print_macro`.
```
I think the idea was that we are consistent with the docs and simplify copy-pasting between docs and comments. But it is true that intra-docs links may be broken anyway, i.e. not having them checked does not help. Since only have a few instances of this, yeah, let's remove them. (We could add a mention of this in `Documentation/rust/coding-guidelines.rst`).

(It would be ideal if `rustdoc` gained eventually the possibility of checking these links + create actual links in the source view form of the docs.)

vobst replied:
Should the removal be in the same commit as the doc update?

> (It would be ideal if rustdoc gained eventually the possibility of checking these links + create actual links in the source view form of the docs.)

Indeed, that would be nice

ojeda replied:
Sorry, which doc update? Do you mean the coding guidelines one I suggested? Yeah, it could be the same commit (or not). But if you are going to do the guidelines change, then the important bit is that part, since we are adding a rule, so I would probably submit that change (actual removal and guideline change) independently of this patch series.

vobst replied:
Yes, I meant the coding guidelines change (there is too much usage of the word doc here).

It would be a separate two-patch patch set then; One patch for the removals and one patch for the coding guidelines change. One could add it in the section on documentation, which is where the doclinks are introduced.

```diff
diff --git a/Documentation/rust/coding-guidelines.rst b/Documentation/rust/coding-guidelines.rst
index 05542840b..4425d391e 100644
--- a/Documentation/rust/coding-guidelines.rst
+++ b/Documentation/rust/coding-guidelines.rst
@@ -164,6 +164,9 @@ in the kernel:
   - Rust items (functions, types, constants...) must be linked appropriately
     (``rustdoc`` will create a link automatically).

+    Please note that ``rustdoc`` does not process links in regular comments.
+    Therefore, the special link syntax should not be used there.
+
   - Any ``unsafe`` block must be preceded by a ``// SAFETY:`` comment
     describing why the code inside is sound.
```

ojeda commented on rust/kernel/types.rs:
```diff
@@ -261,7 +261,7 @@ impl<T> Opaque<T> {
 }
 
-/// Types that are _always_ reference counted.
+/// Types that are _always_ reference-counted.
```
Not sure about this one (the commit message does not mention it either).

Also, for the commit message: it would help mentioning how much widely used it is on each side (i.e. a figure), e.g. grepping I see 350+ vs. 20+.

vobst replied:
For the short form the situation is pretty clear:

```console
$ rg '(\*|//).*?\srefcount(|ed)[\s,.]' | wc -l
1605
$ rg '(\*|//).*?\sref-count(|ed)[\s,.]' | wc -l
43
```

For the long form the situation is less clear:

```console
$ rg '(\*|//).*?\sreference counted[\s,.]' | wc -l
44
$ rg '(\*|//).*?\sreference-counted[\s,.]' | wc -l
20
```

Here's the rust repo for comparison:

```console
$ rg '(\*|//).*?\sreference counted[\s,.]' | wc -l
12
$ rg '(\*|//).*?\sreference-counted[\s,.]' | wc -l
47
```

But I guess for the long version its rather a question of English grammar that statistics. Looking at it a bit longer, the above fix seems wrong as the compound modifier is after the noun. In the cases where it is before the noun using the hyphen seems to be correct. Would be glad if a native speaker could confirm, then I'd adjust this commit accordingly (and mention it in the commit message).

ojeda replied:
Thanks! Yeah, in that case, for the long one, I would avoid changing it unless someone confirms it should be changed.

bjorn3 commented:
> The commits in github are from an unknown profile.

If you actually look at the commits you will see that the author name and email are set correctly. They probably didn't add the email they used to their github account, so github doesn't show the associated github account as author.

ojeda reviewed (commented):
> running `make rusttest` to (hopefully) run the doctests

`rusttest` runs the few `#[test]`s we have. For doctests, you want to use KUnit (make sure you have `CONFIG_RUST_KERNEL_DOCTESTS=y`). Please see https://docs.kernel.org/next/rust/general-information.html#testing.

> is the splitting into individual commits appropriate?

Definitely appropriate!

I think you did commits by area in some cases, which is fine, but for trivial-enough things that apply to more places, you can do a single commit for everything. For instance `rust: init: fix multiple typos in documentation` could be `rust: kernel: fix multiple typos in documentation` and, given its "generic" title, it could take typos from some of the other commits perhaps (e.g. the one from `rust: workqueue: fix trivial typo in docs of __enqueue`).

Also, in `rust: workqueue: add doclinks`, you also remove others. So the title should probably be different (e.g. "improve"?). Or you can keep the title and do the removals in another commit (and, again, ideally applied to more than just `workqueue`).

> are the commit messages OK?

Commits need to be signed (in the `Signed-off-by` sense). Please see https://docs.kernel.org/process/submitting-patches.html on what that implies (Developer's Certificate of Origin).

Also, you can use Markdown in commit messages if you wish (in the Rust subsystem we typically do, but different people in the kernel use different styles).

> should I make one or two patch sets?

For something like this, it is fine to group them, since they are related, small enough and can likely all go in together.

vobst commented:
> `rusttest` only runs the few `#[test]`s we have. For doctests, you want to use KUnit (make sure you have `CONFIG_RUST_KERNEL_DOCTESTS=y`). Please see https://docs.kernel.org/next/rust/general-information.html#testing.
> 

Ah, ok, I was already suspecting that this wasn't the right thing. Will do that.

> I think you did commits by area in some cases, which is fine, but for trivial-enough things that apply to more places, you can do a single commit for everything. For instance `rust: init: fix multiple typos in documentation` could be `rust: kernel: fix multiple typos in documentation` and, given its "generic" title, it could take typos from some of the other commits perhaps (e.g. the one from `rust: workqueue: fix trivial typo in docs of __enqueue`).
> 

Sure, I'll squash those trivial typofixes.

> Also, in `rust: workqueue: add doclinks`, you also remove others. So the title should probably be different (e.g. "improve"?). Or you can keep the title and do the removals in another commit (and, again, ideally applied to more than just `workqueue`).
>

Afaik this commit should only add doc links. The removals are cases where it is not necessary to give the explicit link target as rustdoc can figure it out itself. They could go into their own commit though.

> Also, you can use Markdown in commit messages if you wish (in the Rust subsystem we typically do, but different people in the kernel use different styles).
> 

Nice, I prefer Markdown as well.

BennoLossin commented on rust/kernel/init.rs:
````diff
@@ -235,7 +235,7 @@ pub mod macros;
 /// # Examples
 ///
-/// ```rust
+/// ```
````
I know that `rust` is the default, but I still put that there, because it makes it easier to copy paste the example to a different place. Not sure if we have a standard regarding this though...

ojeda replied:
That makes sense. On one hand, it adds some boilerplate; on the other hand, it is convenient when copying it to an standalone Markdown document.

In the `kernel` crate, the only `rust`-using ones are in `init`. In the Rust `library`/`compiler` folders, I see the majority of the code does not use it (an order of magnitude difference or more).

I guess we can poll / ask others for this one, e.g. how often they copy into another document.

vobst replied:
Oh, I did not think of the copy-paste use case. I just had in mind that people who are not used to Rust and Rustdoc might be confused ("Do I have to add `rust` or not?"), and its best if the kernel crate is consistent about it. If you come to the conclusion that its too opinionated I will just drop it.

vobst replied:
One thing that I left out of the patch set are doclinks to the documented item itself, e.g., `Arc` links back to itself three times in its description. Its a bit redundant and e.g. the library is not doing that per its guidelines. If that has not been discussed it might be something to think about.

vobst commented:
@charmitro @bjorn3 Yes, I did not add that email address to my GH account.

vobst commented:
I'll include the changelog between this RFC and the v1 I submit to the mailing list below to aid the reviewers that already gave feedback here:
- squashed trivial typo fixes into a single commit; In particular I squashed
```
11ee461ad rust: workqueue: fix trivial typo in docs of __enqueue
e3f961290 rust: spinlock: fix typo in 'unlock' SAFETY comment
ec6a49930 rust: str: fix trivial typo in bytes_written docs
b0a2c7a70 rust: ioctl: fix indefinite articles in doc comments
fecc9b554 rust: init: fix multiple typos in documentation
c95a3d882 rust: allocator: fix indentation in comment
```
- now using Markdown to highlight code fragments in commit messages
- dropped `d3cb9b509 rust: remove 'rust' annotation on code blocks in docs`
- dropped `a2f0cea40 rust: remove doclinks in regular comments` in favor of a separate patch set that also touches the coding guidelines
- remove changes to long form of refcount and include statistics in commit message (`rust: unify spelling of refcount in docs`)
- `rust: workqueue: add doclinks` is now `rust: kernel: add doclinks` and adds doclinks across the kernel crate; I also moved the parts that remove unneeded explicit doc link targets and introduce new links with the "code" html tag into their own commits
- `rust: mark code elements in comments with backticks` now includes the `ioctl()`
- `rust: kernel: add doclinks` now includes `Sized` in the workqueue module
- Added "Signed-of-by" to all commits.
- There is one new commit `rust: locked_by: shorten doclink preview`. It changes two link previews from `super::type name` to `type name`.

This time I also properly ran the doctests.

Edit: the v1 tree is [here](https://github.com/vobst/rust4lx/tree/doc-fixes-v1)
